Initial implementation of WebGLQueries

mmatyas opened the pull request:
This patch adds initial support for WeGLQueries. Most related WebGL functions and objects ([1]) are implemented; what's still missing is:

- syncing the result of `getQueryParameter` with the event loop
- `EXT_disjoint_timer_query_webgl2` support

[1]: https://www.khronos.org/registry/webgl/specs/latest/2.0/#3.7.12

<!-- Please describe your changes on the following line: -->
cc @jdm @zakorgy 

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/24178)
<!-- Reviewable:end -->

mmatyas committed:
Initial implementation of WebGLQueries

This patch adds initial support for WeGLQueries. Most related WebGL
functions and objects are implemented [1]. What's still missing is
the `EXT_disjoint_timer_query_webgl2` support.

[1]: https://www.khronos.org/registry/webgl/specs/latest/2.0/#3.7.12

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/webglrenderingcontext.rs, components/script/dom/webidls/WebGL2RenderingContext.webidl, components/script/dom/webidls/WebGLQuery.webidl, components/script/dom/mod.rs, components/script/dom/bindings/trace.rs and 2 more
 * @jgraham: tests/wpt/webgl/meta/conformance2/query/occlusion-query.html.ini, tests/wpt/webgl/meta/conformance2/query/query.html.ini
 * @kichjang: components/script/dom/webglrenderingcontext.rs, components/script/dom/webidls/WebGL2RenderingContext.webidl, components/script/dom/webidls/WebGLQuery.webidl, components/script/dom/mod.rs, components/script/dom/bindings/trace.rs and 2 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

mmatyas commented:
Note: the patch doesn't pass `test-tidy` at the moment, because in my opinion this:

```rust
match target {
    constants::ANY_SAMPLES_PASSED |
    constants::ANY_SAMPLES_PASSED_CONSERVATIVE => {
        &self.occlusion_query
    },
    constants::TRANSFORM_FEEDBACK_PRIMITIVES_WRITTEN => {
        &self.primitives_query
    },
```

is more readable than this:

```rust
match target {
    constants::ANY_SAMPLES_PASSED | constants::ANY_SAMPLES_PASSED_CONSERVATIVE => {
        &self.occlusion_query
    },
    constants::TRANSFORM_FEEDBACK_PRIMITIVES_WRITTEN => &self.primitives_query,
```

Of course, I can apply `test-tidy`'s changes in case you disagree.


Also I'm pretty sure many Option check could be simplified with some creative use of `and_then` and co., I'd be happy to hear your thoughs about them.

jdm commented on components/script/dom/webgl2renderingcontext.rs:
```diff
@@ -1044,6 +1051,149 @@ impl WebGL2RenderingContextMethods for WebGL2RenderingContext {
+        );
+        match query.get_parameter(&self.base, pname) {
+            Ok(value) => match pname {
+                constants::QUERY_RESULT => UInt32Value(value),
```
To get the event loop integration right, I recommend this strategy:
* add `Option<u32>` and `Option<bool>` members for storing cached versions of these results
* if both these fields are None, get the query parameter from the GL thread and store the result in the appropriate field, and return the result
* if either field is Some, get the query parameter from the GL thread, queue a task that will update the cached values, and return the current cached value

mmatyas replied:
I see, but what do you mean exactly by queuing a task? The query results should not change until the next time we're in the event loop; do we have a way for scheduling a callback to run at that time? Or you mean connect it somehow to the DOM's global scheduler?

jdm replied:
I mean using the `task!` macro to declare a closure that will run asynchronously on the script thread, and using the task source API to enqueue it to run later. [This code](https://github.com/servo/servo/blob/e431a7baf712182f10782d0264399917a185c128/components/script/dom/storage.rs#L210-L233) is one example of many in the script/ code of queueing a task.

jdm reviewed (changes requested):
The rest of the changes look good!

jdm commented:
As for the question about tidy, you can add a `#[cfg(rustfmt_skip)` annotation if you would prefer.

mmatyas commented:
Ok, I've updated the patch with the task queue (thanks!), and I've also added a Drop implementation. The timing tests now pass, but the actual query results might fail (`wtu.drawUnitQuad` produces a crash on my PC, unrelated to this patch). I wonder if it passes on the CI.

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit bea268223e6df7bca416502b610c2c532c73fd04 with merge d4fb458534088476feeca2f599b09f3640c1c35b...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/13524)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #24242) made this pull request unmergeable. Please resolve the merge conflicts.

mmatyas commented:
Rebased to master. I've also improved the `Drop` impl making it support the fallible case. In the commit message I've removed the note about the missing event loop sync implementation, as it now works too.

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 676a52d5c6f97fbc82e0456d313d78830d4befc3 with merge dde302e710dbeb3796ef5e58486bac140fa63ff0...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/13498)

jdm commented:
You can ignore the /_webgl/conformance/ogles/GL/asin/asin_001_to_006.html crash; interfaces.html will need to be updated to include the new WebGLQuery interface.

mmatyas commented:
Ok, updated with the test results. Actually we could pass even more if we'd run `test-wpt` with `--pref dom.webgl2.enabled`; do you plan to enable this flag on the builders?

jdm commented:
It is already enabled for the webgl2 conformance tests by https://github.com/servo/servo/blob/master/tests/wpt/webgl/meta/conformance2/__dir__.ini.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #24314) made this pull request unmergeable. Please resolve the merge conflicts.

mmatyas commented:
> It is already enabled for the webgl2 conformance tests by https://github.com/servo/servo/blob/master/tests/wpt/webgl/meta/conformance2/__dir__.ini.

Ah, I see, thanks.

> What happens if the query result is still not available when update_results runs? It looks like this will only ever check once.

If the result is not available, `query_result_available` will be again set to `None`, allowing `get_parameter` to queue a task again.

mmatyas commented:
Ok, rebased to master. There's this interesting test after the recent WPT update, `conformance2/context/methods-2.html`, which tests for every WebGL 1/2 function's existence. It would of course fail for us, yet it seems it's expected to pass; I wonder what's up with that.

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 86d971d1a21772920ca711a3531e2c04bcf60766 with merge 0252b72f28db00691b0fa80b4cd4da0f40e45eeb...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/13539)

jdm commented:
I see what you mean:
```
  ▶ Unexpected subtest result in /_webgl/conformance2/context/methods-2.html:
  │ FAIL [expected PASS] WebGL test #43: Property either does not exist or is not a function: createSampler
  │   → assert_true: Property either does not exist or is not a function: createSampler expected true got false
  │ FAIL [expected PASS] WebGL test #44: Property either does not exist or is not a function: deleteSampler
  │   → assert_true: Property either does not exist or is not a function: deleteSampler expected true got false
  │ FAIL [expected PASS] WebGL test #45: Property either does not exist or is not a function: isSampler
  │   → assert_true: Property either does not exist or is not a function: isSampler expected true got false
  │ FAIL [expected PASS] WebGL test #46: Property either does not exist or is not a function: bindSampler
  │   → assert_true: Property either does not exist or is not a function: bindSampler expected true got false
  │ FAIL [expected PASS] WebGL test #47: Property either does not exist or is not a function: samplerParameteri
  │   → assert_true: Property either does not exist or is not a function: samplerParameteri expected true got false
  │ FAIL [expected PASS] WebGL test #48: Property either does not exist or is not a function: samplerParameterf
  │   → assert_true: Property either does not exist or is not a function: samplerParameterf expected true got false
  │ FAIL [expected PASS] WebGL test #49: Property either does not exist or is not a function: getSamplerParameter
  │   → assert_true: Property either does not exist or is not a function: getSamplerParameter expected true got false
  │ FAIL [expected PASS] WebGL test #50: Property either does not exist or is not a function: fenceSync
  │   → assert_true: Property either does not exist or is not a function: fenceSync expected true got false
  │ FAIL [expected PASS] WebGL test #51: Property either does not exist or is not a function: isSync
  │   → assert_true: Property either does not exist or is not a function: isSync expected true got false
  │ FAIL [expected PASS] WebGL test #52: Property either does not exist or is not a function: deleteSync
  │   → assert_true: Property either does not exist or is not a function: deleteSync expected true got false
  │ FAIL [expected PASS] WebGL test #53: Property either does not exist or is not a function: clientWaitSync
  │   → assert_true: Property either does not exist or is not a function: clientWaitSync expected true got false
  │ FAIL [expected PASS] WebGL test #54: Property either does not exist or is not a function: waitSync
  │   → assert_true: Property either does not exist or is not a function: waitSync expected true got false
  │ FAIL [expected PASS] WebGL test #55: Property either does not exist or is not a function: getSyncParameter
  │   → assert_true: Property either does not exist or is not a function: getSyncParameter expected true got false
  │ FAIL [expected PASS] WebGL test #56: Property either does not exist or is not a function: createTransformFeedback
  │   → assert_true: Property either does not exist or is not a function: createTransformFeedback expected true got false
  │ FAIL [expected PASS] WebGL test #57: Property either does not exist or is not a function: deleteTransformFeedback
  │   → assert_true: Property either does not exist or is not a function: deleteTransformFeedback expected true got false
  │ FAIL [expected PASS] WebGL test #58: Property either does not exist or is not a function: isTransformFeedback
  │   → assert_true: Property either does not exist or is not a function: isTransformFeedback expected true got false
  │ FAIL [expected PASS] WebGL test #59: Property either does not exist or is not a function: bindTransformFeedback
  │   → assert_true: Property either does not exist or is not a function: bindTransformFeedback expected true got false
  │ FAIL [expected PASS] WebGL test #60: Property either does not exist or is not a function: beginTransformFeedback
  │   → assert_true: Property either does not exist or is not a function: beginTransformFeedback expected true got false
  │ FAIL [expected PASS] WebGL test #61: Property either does not exist or is not a function: endTransformFeedback
  │   → assert_true: Property either does not exist or is not a function: endTransformFeedback expected true got false
  │ FAIL [expected PASS] WebGL test #62: Property either does not exist or is not a function: transformFeedbackVaryings
  │   → assert_true: Property either does not exist or is not a function: transformFeedbackVaryings expected true got false
  │ FAIL [expected PASS] WebGL test #63: Property either does not exist or is not a function: getTransformFeedbackVarying
  │   → assert_true: Property either does not exist or is not a function: getTransformFeedbackVarying expected true got false
  │ FAIL [expected PASS] WebGL test #64: Property either does not exist or is not a function: pauseTransformFeedback
  │   → assert_true: Property either does not exist or is not a function: pauseTransformFeedback expected true got false
  │ FAIL [expected PASS] WebGL test #65: Property either does not exist or is not a function: resumeTransformFeedback
  │   → assert_true: Property either does not exist or is not a function: resumeTransformFeedback expected true got false
  │ FAIL [expected PASS] WebGL test #66: Property either does not exist or is not a function: bindBufferBase
  │   → assert_true: Property either does not exist or is not a function: bindBufferBase expected true got false
  │ FAIL [expected PASS] WebGL test #67: Property either does not exist or is not a function: bindBufferRange
  │   → assert_true: Property either does not exist or is not a function: bindBufferRange expected true got false
  │ FAIL [expected PASS] WebGL test #68: Property either does not exist or is not a function: getIndexedParameter
  │   → assert_true: Property either does not exist or is not a function: getIndexedParameter expected true got false
  │ FAIL [expected PASS] WebGL test #69: Property either does not exist or is not a function: getUniformIndices
  │   → assert_true: Property either does not exist or is not a function: getUniformIndices expected true got false
  │ FAIL [expected PASS] WebGL test #70: Property either does not exist or is not a function: getActiveUniforms
  │   → assert_true: Property either does not exist or is not a function: getActiveUniforms expected true got false
  │ FAIL [expected PASS] WebGL test #71: Property either does not exist or is not a function: getUniformBlockIndex
  │   → assert_true: Property either does not exist or is not a function: getUniformBlockIndex expected true got false
  │ FAIL [expected PASS] WebGL test #72: Property either does not exist or is not a function: getActiveUniformBlockParameter
  │   → assert_true: Property either does not exist or is not a function: getActiveUniformBlockParameter expected true got false
  │ FAIL [expected PASS] WebGL test #73: Property either does not exist or is not a function: getActiveUniformBlockName
  │   → assert_true: Property either does not exist or is not a function: getActiveUniformBlockName expected true got false
  │ FAIL [expected PASS] WebGL test #74: Property either does not exist or is not a function: uniformBlockBinding
  │   → assert_true: Property either does not exist or is not a function: uniformBlockBinding expected true got false
  │ FAIL [expected PASS] WebGL test #75: Property either does not exist or is not a function: createVertexArray
  │   → assert_true: Property either does not exist or is not a function: createVertexArray expected true got false
  │ FAIL [expected PASS] WebGL test #76: Property either does not exist or is not a function: deleteVertexArray
  │   → assert_true: Property either does not exist or is not a function: deleteVertexArray expected true got false
  │ FAIL [expected PASS] WebGL test #77: Property either does not exist or is not a function: isVertexArray
  │   → assert_true: Property either does not exist or is not a function: isVertexArray expected true got false
  │ FAIL [expected PASS] WebGL test #78: Property either does not exist or is not a function: bindVertexArray
  │   → assert_true: Property either does not exist or is not a function: bindVertexArray expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  │ test@http://web-platform.test:8000/resources/testharness.js:544:30
  │ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:7
  │ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:5
  │ assertFunction@http://web-platform.test:8000/_webgl/conformance2/context/methods-2.html:286:7
  └ @http://web-platform.test:8000/_webgl/conformance2/context/methods-2.html:304:11
```

jdm commented on components/script/dom/webidls/WebGLQuery.webidl:
```diff
@@ -0,0 +1,11 @@
+//
+
+[Exposed=Window]
+interface WebGLQuery : WebGLObject {
```
This needs to be hidden behind the webgl2 preference: https://github.com/servo/servo/blob/86d971d1a21772920ca711a3531e2c04bcf60766/components/script/dom/webidls/WebGL2RenderingContext.webidl#L583

jdm commented:
Here's the reason for those failures - the test harness automatically adds a #N prefix to each test message in [this code](https://searchfox.org/mozilla-central/source/dom/canvas/test/webgl-conf/checkout/conformance2/context/methods-2.html#284-290). Since it only calls testFailed if a method is missing, it now executes testFailed fewer times than previously since there are new methods present in the interface, to the test messages contain different #N prefixes than before. You should delete tests/wpt/webgl/meta/conformance2/context/methods-2.html.ini and regenerate it in order to update the expected error messages.

jdm commented:
Also some other interface test expectations needs updating:
```
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "WebGLQuery interface: existence and properties of interface object", 
    "test": "/webgl/idlharness.any.html", 
    "line": 332663, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "WebGLQuery interface object length", 
    "test": "/webgl/idlharness.any.html", 
    "line": 332664, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "WebGLQuery interface object name", 
    "test": "/webgl/idlharness.any.html", 
    "line": 332665, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "WebGLQuery interface: existence and properties of interface prototype object", 
    "test": "/webgl/idlharness.any.html", 
    "line": 332666, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "WebGLQuery interface: existence and properties of interface prototype object's \"constructor\" property", 
    "test": "/webgl/idlharness.any.html", 
    "line": 332667, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "WebGLQuery interface: existence and properties of interface prototype object's @@unscopables property", 
    "test": "/webgl/idlharness.any.html", 
    "line": 332668, 
    "action": "test_result", 
    "expected": "FAIL"
}
```

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit f2e2b3d has been approved by `jdm`

<!-- @bors-servo r=jdm f2e2b3d34b5ed1afa2b937eedee4908cd5d07a05 -->

bors-servo commented:
:hourglass: Testing commit f2e2b3d34b5ed1afa2b937eedee4908cd5d07a05 with merge cfcff8e771b1395f35901abb040049dfa2c6d262...

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit f2e2b3d34b5ed1afa2b937eedee4908cd5d07a05 with merge d3475e5c3f17b89d7c9d3f36a065e56169e39060...

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/13560), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/13622), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/Ovr5q64dR7-Oic3UA_DtvA)
Approved by: jdm
Pushing d3475e5c3f17b89d7c9d3f36a065e56169e39060 to master...
